### PR TITLE
Issue #64, #team2

### PR DESCRIPTION
**Pull request by ghost**
Issue #64,  after user logouts, she will be redirect back to the page she came from.

### CODE REVIEW

File: piplmesh/account/signals.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from django import dispatch
+from django.contrib import messages
+from django.utils.translation import ugettext_lazy as _
+
+user_login = dispatch.Signal(providing_args=["request", "user"])
```
Comment (mitar) [inline piplmesh/account/signals.py]: Constant strings should be in `'`.

Comment (mitar) [inline piplmesh/account/signals.py]: Those signals already exist in `django.contrib.auth.signals`. You can just reuse them?

File: piplmesh/account/views.py [hunk 2]
```diff
@@ -63,4 +65,23 @@ def get(self, request, *args, **kwargs):
...
-            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
\ No newline at end of file
+            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
+
+def logout(request):
+    """
+    After user logouts, redirect her back to the page she came from.
+    """
+    
+    url = request.META.get('HTTP_REFERER')
```
Comment (mitar) [inline piplmesh/account/views.py]: No no, we will not go over `HTTP_REFERER`. It does not work always and especially not when using HTTPS (browsers do not send it with HTTPS).

File: piplmesh/account/views.py [hunk 3]
```diff
@@ -63,4 +65,23 @@ def get(self, request, *args, **kwargs):
...
+    return logout_redirect(request, next_page=url)
+
+def logout_redirect(request, *args, **kwargs):
```
Comment (mitar) [inline piplmesh/account/views.py]: What is wrong with existing `logout_then_login` view?

File: piplmesh/account/views.py [hunk 4]
```diff
@@ -63,4 +65,23 @@ def get(self, request, *args, **kwargs):
...
+    """
+    Logs out the user and redirects her to the log-in page or elsewhere, as specified.
+    """    
+    
+    kwargs.setdefault('redirect_field_name')
```
Comment (mitar) [inline piplmesh/account/views.py]: Why is this needed?

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 5]
```diff
@@ -282,4 +282,14 @@ ul.errorlist {
     line-height: normal;
     border-bottom: dotted 1px #999;
     cursor: pointer;
+}
+
+/* ------------------- MESSAGES ------------------- */
+.messages{
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: Please format consistently. Use spaces before `{`. Spaces after `:`. Indent properly.

File: piplmesh/frontend/templates/base.html [hunk 6]
```diff
@@ -13,6 +13,7 @@
         {% block css %}{% endblock %}
     </head>
     <body>
-        {% block body %}{% endblock %}
+        {% block body %}
```
Comment (mitar) [inline piplmesh/frontend/templates/base.html]: Why changing this?

File: piplmesh/frontend/templates/messages.html [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{% if messages %}
+<div id="messages">
```
Comment (mitar) [inline piplmesh/frontend/templates/messages.html]: Please indent the same as elsewhere.

File: piplmesh/account/views.py (hunk 3 above)
Comment (ghost) [inline piplmesh/account/views.py]: Nothing. I missed this view. :)

File: piplmesh/frontend/templates/base.html (hunk 6 above)
Comment (ghost) [inline piplmesh/frontend/templates/base.html]: Mistake.

File: piplmesh/frontend/static/piplmesh/css/style.css (hunk 5 above)
Comment (ghost) [inline piplmesh/frontend/static/piplmesh/css/style.css]: Corrected.

File: piplmesh/account/signals.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+from django import dispatch
+from django.contrib.auth import signals
+from django.contrib import messages
```
Comment (mitar) [inline piplmesh/account/signals.py:2]: Import order.

File: piplmesh/account/views.py [hunk 9]
```diff
@@ -4,10 +4,12 @@
 from django.conf import settings
 from django.contrib import auth
 from django.core import urlresolvers
+from django.contrib.auth import views as auth_views
 from django.views import generic as generic_views
 from django.views.generic import simple, edit as edit_views
 
 from piplmesh.account import forms
+from piplmesh.account import signals
```
Comment (mitar) [inline piplmesh/account/views.py]: You can combine those two imports.

File: piplmesh/account/views.py [hunk 10]
```diff
@@ -4,10 +4,12 @@
 from django.conf import settings
 from django.contrib import auth
 from django.core import urlresolvers
+from django.contrib.auth import views as auth_views
```
Comment (mitar) [inline piplmesh/account/views.py:6]: Order of imports.

File: piplmesh/frontend/templates/messages.html (hunk 7 above)
Comment (mitar) [inline piplmesh/frontend/templates/messages.html]: TODO.

File: piplmesh/frontend/static/piplmesh/css/style.css [hunk 11]
```diff
@@ -282,4 +282,14 @@ ul.errorlist {
     line-height: normal;
     border-bottom: dotted 1px #999;
     cursor: pointer;
+}
+
+/* ------------------- MESSAGES ------------------- */
+.messages {
+	border: solid 1px #DEDEDE;
```
Comment (mitar) [inline piplmesh/frontend/static/piplmesh/css/style.css]: Indentation?

File: piplmesh/account/views.py (hunk 2 above)
Comment (mitar) [inline piplmesh/account/views.py]: No, this is not good.

Comment (ghost) [inline piplmesh/account/views.py]: Corrected, but I don't know how to get query string from previous url..?

File: piplmesh/frontend/templates/messages.html (hunk 7 above)
Comment (ghost) [inline piplmesh/frontend/templates/messages.html]: Done.

File: piplmesh/account/views.py [hunk 12]
```diff
@@ -63,4 +64,12 @@ def get(self, request, *args, **kwargs):
...
-            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
\ No newline at end of file
+            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
+
+def logout(request):
+    """
+    After user logouts, redirect her back to the page she came from.
+    """
+    
+    url = request.REQUEST.get('next', '')
```
Comment (mitar) [inline piplmesh/account/views.py]: Why `''` and not `None`? Is second argument really necessary?

File: piplmesh/frontend/templates/messages.html [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{% if messages %}
```
Comment (mitar) [inline piplmesh/frontend/templates/messages.html:1]: Indendation?

File: piplmesh/frontend/templates/header.html [hunk 14]
```diff
@@ -28,7 +29,7 @@
                 <li class="image"><img src="{% facebook_picture user.username "square" %}" alt="{% trans "User picture" %}"/></li>
                 <li class="username">{{ user.username }}</li>
                 <li class="last">
-                    <form method="post" action="{% url logout %}" id="logout_form">
+                    <form method="post" action="{% url logout %}?next={{ request.path }}" id="logout_form">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: You could move `next` to hidden field and then access it simply through POST?

File: piplmesh/account/views.py (hunk 12 above)
Comment (mitar) [inline piplmesh/account/views.py]: Can you check if it is POST request we are in? And otherwise throw a 403 error?

Comment (mitar) [inline piplmesh/account/views.py]: Why not `request.POST.get('next')`?

Comment (mitar) [inline piplmesh/account/views.py]: And don't hard-code `next`, use constant variable, `REDIRECT_FIELD_NAME`!

File: piplmesh/frontend/templates/messages.html (hunk 13 above)
Comment (ghost) [inline piplmesh/frontend/templates/messages.html:1]: I hope that is OK now...

File: piplmesh/frontend/templates/header.html (hunk 14 above)
Comment (ghost) [inline piplmesh/frontend/templates/header.html]: Done!

File: piplmesh/account/views.py (hunk 12 above)
Comment (ghost) [inline piplmesh/account/views.py]: Done! I throw default 403 error, is this OK?

File: piplmesh/account/views.py [hunk 15]
```diff
@@ -63,4 +65,15 @@ def get(self, request, *args, **kwargs):
...
-            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
\ No newline at end of file
+            return super(FacebookCallbackView, self).get(request, *args, **kwargs)
+
+def logout(request):
+    """
+    After user logouts, redirect her back to the page she came from.
+    """
+    
+    if request.method == "POST":
```
Comment (mitar) [inline piplmesh/account/views.py]: Constant string! ;-)

File: piplmesh/account/views.py [hunk 16]
```diff
@@ -63,4 +65,15 @@ def get(self, request, *args, **kwargs):
...
+        url = request.POST.get(auth.REDIRECT_FIELD_NAME)
+        return auth_views.logout_then_login(request, url)
+    else:
+        return HttpResponse(status=403)
```
Comment (mitar) [inline piplmesh/account/views.py]: Just raise `PermissionDenied`.

File: piplmesh/account/views.py [hunk 17]
```diff
@@ -3,11 +3,13 @@
 from django import http
 from django.conf import settings
 from django.contrib import auth
+from django.contrib.auth import views as auth_views
 from django.core import urlresolvers
+from django.http import HttpResponse
```
Comment (mitar) [inline piplmesh/account/views.py]: Import only modules!

File: piplmesh/frontend/templates/header.html [hunk 18]
```diff
@@ -31,6 +32,7 @@
                     <form method="post" action="{% url logout %}" id="logout_form">
                         {% csrf_token %}
                         <div>
+                            <input type="hidden" name="next" value="{{ request.path }}" />
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Don't hardcode next, but get it from the code. (We have a global context processor to pass such values to templates.)

Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Indentation.

File: piplmesh/frontend/templates/messages.html (hunk 13 above)
Comment (mitar) [inline piplmesh/frontend/templates/messages.html:1]: Me too. :-)

File: piplmesh/urls.py [hunk 19]
```diff
@@ -20,7 +20,7 @@
     # Registration, login, logout
     url(r'^register/$', account_views.RegistrationView.as_view(), name='registration'),
     url(r'^login/$', 'django.contrib.auth.views.login', {'template_name': 'login.html'}, name='login'),
```
Comment (mitar) [inline piplmesh/urls.py]: And if we are here. Can you just put the template into its default location and get rid of this parameter?

File: piplmesh/urls.py [hunk 20]
```diff
@@ -20,7 +20,7 @@
...
-    url(r'^logout/$', 'django.contrib.auth.views.logout', {'template_name': 'logout.html'}, name='logout'),
+    url(r'^logout/$', account_views.logout, name='logout'),
     # Facebook
     url(r'^facebook/login/$', account_views.FacebookLoginView.as_view(), name='facebook_login'),
```
Comment (mitar) [inline piplmesh/urls.py:25]: Probably you should also display message for Facebook login?

Comment (ghost) [inline piplmesh/urls.py:25]: Message is also displayed when user logs in with Facebook.

File: piplmesh/urls.py (hunk 19 above)
Comment (ghost) [inline piplmesh/urls.py]: Done!

File: piplmesh/account/views.py (hunk 16 above)
Comment (ghost) [inline piplmesh/account/views.py]: Done.

File: piplmesh/account/views.py (hunk 15 above)
Comment (ghost) [inline piplmesh/account/views.py]: Fixed.

File: piplmesh/account/templates/registration/login.html [hunk 21]
```diff
@@ -0,0 +1,31 @@
```
Comment (mitar) [inline piplmesh/account/templates/registration/login.html]: Shouldn't this be a move?

File: piplmesh/account/templates/registration/registration.html [hunk 22]
```diff
@@ -0,0 +1,34 @@
```
Comment (mitar) [inline piplmesh/account/templates/registration/registration.html]: Shouldn't this be a move?

File: piplmesh/frontend/context_processors.py [hunk 23]
```diff
@@ -7,4 +7,5 @@ def global_vars(request):
 
     return {
         'SEARCH_ENGINE_UNIQUE_ID': settings.SEARCH_ENGINE_UNIQUE_ID,
+        'REDIRECT_FIELD_NAME': 'next',
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: Noooo. You have this value in `contrib.django.auth.REDIRECT_FIELD_NAME`!

File: piplmesh/account/views.py [hunk 24]
```diff
@@ -63,4 +64,15 @@ def get(self, request, *args, **kwargs):
...
+def logout(request):
+    """
+    After user logouts, redirect her back to the page she came from.
+    """
+    
+    if request.method == 'POST':
+        url = request.POST.get(auth.REDIRECT_FIELD_NAME)
+        return auth_views.logout_then_login(request, url)
+    else:
+        raise exceptions.PermissionDenied()
```
Comment (mitar) [inline piplmesh/account/views.py]: I think just `raise exceptions.PermissionDenied` also works? (Without `()`.)

File: piplmesh/frontend/templates/header.html [hunk 25]
```diff
@@ -31,6 +32,7 @@
                     <form method="post" action="{% url logout %}" id="logout_form">
                         {% csrf_token %}
                         <div>
+                            <input type="hidden" name="{{ REDIRECT_FIELD_NAME }}" value="{{ request.path }}" />
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: OK. Better. But for the value you will want a bit better. Check how it is done in nodewatcher.

```
{% with next|default:context_processor_next|default:request.get_full_path as next_url %}
```

So we allow three values:
- from context
- from context_processor
- from request

(In this order.)

File: piplmesh/account/views.py (hunk 24 above)
Comment (ghost) [inline piplmesh/account/views.py]: Yes this also works without '()'.

File: piplmesh/frontend/templates/header.html (hunk 25 above)
Comment (ghost) [inline piplmesh/frontend/templates/header.html]: I think it is OK now.

File: piplmesh/frontend/context_processors.py (hunk 23 above)
Comment (ghost) [inline piplmesh/frontend/context_processors.py]: fixed... :)

File: piplmesh/account/signals.py [hunk 26]
```diff
@@ -0,0 +1,22 @@
+from django import dispatch
+from django.contrib import messages
+from django.contrib.auth import signals
+from django.utils.translation import ugettext_lazy as _
+
+def user_login_message(sender, request, user, **kwargs):
+    """
+    Show success login message.
```
Comment (mitar) [inline piplmesh/account/signals.py]: _Show_ -> _Shows_

File: piplmesh/account/signals.py [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+    """
+    
+    messages.success(request, _("You have been successfully logged in."), fail_silently=True)
+
+signals.user_logged_in.connect(user_login_message, dispatch_uid=__name__ + '.user_login_message')
+
+def user_logout_message(sender, request, user, **kwargs):
+    """
+    Show success logout message.
```
Comment (mitar) [inline piplmesh/account/signals.py]: _Show_ -> _Shows_

File: piplmesh/frontend/templates/header.html [hunk 28]
```diff
@@ -1,4 +1,5 @@
 {% load i18n facebook %}
+{% with next|default:REDIRECT_TO|default:request.get_full_path as next_url %}
 <div id="container">
```
Comment (mitar) [inline piplmesh/frontend/templates/header.html]: Can you indent the whole block?

File: piplmesh/frontend/context_processors.py [hunk 29]
```diff
@@ -7,4 +8,6 @@ def global_vars(request):
 
     return {
         'SEARCH_ENGINE_UNIQUE_ID': settings.SEARCH_ENGINE_UNIQUE_ID,
+        'REDIRECT_FIELD_NAME': auth.REDIRECT_FIELD_NAME,
+        'REDIRECT_TO': request.POST.get(auth.REDIRECT_FIELD_NAME),
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: This is not really a constant, so it should be lower case.

Comment (ghost) [inline piplmesh/frontend/context_processors.py]: Ok.

File: piplmesh/frontend/templates/header.html (hunk 28 above)
Comment (ghost) [inline piplmesh/frontend/templates/header.html]: Indented.

File: piplmesh/frontend/context_processors.py [hunk 30]
```diff
@@ -7,4 +8,6 @@ def global_vars(request):
 
     return {
         'SEARCH_ENGINE_UNIQUE_ID': settings.SEARCH_ENGINE_UNIQUE_ID,
+        'REDIRECT_FIELD_NAME': auth.REDIRECT_FIELD_NAME,
+        'redirect_to': request.POST.get(auth.REDIRECT_FIELD_NAME),
```
Comment (mitar) [inline piplmesh/frontend/context_processors.py]: Use [REQUEST](https://docs.djangoproject.com/en/dev/ref/request-response/#django.http.HttpRequest.REQUEST):

```
'context_processor_next' : request.REQUEST.get(auth.REDIRECT_FIELD_NAME),
```

Or:

```
'context_processor_next' : request.POST.get(auth.REDIRECT_FIELD_NAME, request.GET.get(auth.REDIRECT_FIELD_NAME)),
```

Comment (ghost) [inline piplmesh/frontend/context_processors.py]: Done.
